src/charm.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+# Copyright 2024 Ubuntu
+# See LICENSE file for licensing details.
+
+"""Charm the application."""
+
+import logging
+
+import ops
+import requests
+from charms.tls_certificates_interface.v3.tls_certificates import (
+    CertificateCreationRequestEvent,
+    TLSCertificatesProvidesV3,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class GocertCharm(ops.CharmBase):
+    """Charmed Gocert."""
+
+    def __init__(self, framework: ops.Framework):
+        super().__init__(framework)
+        self.container = self.unit.get_container("gocert")
+        self.tls = TLSCertificatesProvidesV3(self, relationship_name="certificates")
+
+        framework.observe(self.on["gocert"].pebble_ready, self._on_gocert_pebble_ready)
+        framework.observe(self.on["gocert"].pebble_custom_notice, self._on_gocert_notify)
+        framework.observe(self.tls.on.certificate_creation_request, self._on_new_certificate)
+        framework.observe(self.on.config_changed, self._on_config_changed)
+        framework.observe(self.on.collect_unit_status, self._on_collect_status)
+
+    def _on_gocert_pebble_ready(self, event: ops.PebbleReadyEvent):
+        try:
+            self.container.pull("/etc/config/config.yaml")
+        except ops.pebble.PathError:
+            config_file = open("src/configs/config.yaml").read()
+            cert_file = open("src/configs/cert.pem").read()
+            key_file = open("src/configs/key.pem").read()
```

Comment:
I assume this will be changed to use `TlsCertificates` to request a certificate in the next versions, right? Do you think it would make sense to include that in this PR instead?

Reply:
most likely either that or user secrets or both. Either way i think it's large enough to warrant its own pr.

Reply:
That's ok with me, resolving